File: problem-11/c/main.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

uint64_t max_int(uint64_t a, uint64_t b) {
	if (a < b) return b;
	return a;
}

int read_integers_2d(const char *path, int r, int c, int ***numbers) {
	// Open File
	FILE *fp = NULL;
	if ((fp = fopen(path, "r")) == NULL) {
		printf("fatal - failed to open %s\n", path);
		return 0;
	}

	// Prepare Array
	*numbers = (int**)malloc(r * sizeof(int*));
	for (int i = 0; i < r; i++) {
		(*numbers)[i] = (int *)malloc(c * sizeof(int));
	}

	// Read File
	int count = 0;
	for (int i = 0; i < r; i++) {
		for (int j = 0; j < c; j++) {
			if (fscanf(fp, "%2d", &(*numbers)[i][j]) == 1) {
				count++;
			}
		}
	}

	return count;
}

uint64_t horizontal(int r, int c, int n, int **numbers) {
	uint64_t prod = 1;
	for (int i = 0; i < n; i++) {
		prod *= (uint64_t) numbers[r][c + i];
	}
	return prod;
}

uint64_t vertical(int r, int c, int n, int **numbers) {
	uint64_t prod = 1;
	for (int i = 0; i < n; i++) {
		prod *= (uint64_t) numbers[r + i][c];
	}
	return prod;
}

uint64_t diagonal_1(int r, int c, int n, int **numbers) {
	uint64_t prod = 1;
	for (int i = 0; i < n; i++) {
		prod *= (uint64_t) numbers[r + i][c + i];
	}
	return prod;
}

uint64_t diagonal_2(int r, int c, int n, int **numbers) {
	uint64_t prod = 1;
	for (int i = 0; i < n; i++) {
		prod *= (uint64_t) numbers[r + i][c - i];
	}
	return prod;
}

uint64_t solution(const char *path, int n_rows, int n_cols, int length) {
	// Prepare
	int **numbers = NULL;
	
	// Read File
	int count = 0;
	int num = n_cols * n_rows;
	if ((count = read_integers_2d(path, n_rows, n_cols, &numbers)) != num) {
		printf("fatal - unexpected number of integers read (%d != %d)", count, num);
		return 0;
	}

	// Search
	uint64_t output = 0;
	for (int r = 0; r < n_rows; r++) {
		for (int c = 0; c < n_cols; c++) {
			// Vertical
			if (r < n_rows - length) {
				output = max_int(output, vertical(r, c, length, numbers));
			}

			// Horizontal
			if (c < n_cols - length) {
				output = max_int(output, horizontal(r, c, length, numbers));
			}

			// Diagonal "\"
			if ((r < n_rows - length) && (c < n_cols - length)) {
				output = max_int(output, diagonal_1(r, c, length, numbers));
			}

			// Diagonal "/"
			if ((r < n_rows - length) && (c > length - 1)) {
				output = max_int(output, diagonal_2(r, c, length, numbers));
			}
		}
	}

	// Free Memory
	for (int i = 0; i < n_rows; i++) {
		free(numbers[i]);
	}
	free(numbers);

	return output;
}

int main() {
	char *path = "/home/knight/Documents/github/project-euler/problem-11/input.txt";
	printf("Solution #11: %ld\n", solution(path, 20, 20, 4));
}

File: problem-11/rust/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Error};
use std::cmp;

fn main() {
    let path = "/home/knight/Documents/github/project-euler/problem-11/input.txt";
    println!("Solution #11: {:?}", solution(path, 20, 20, 4).unwrap());
}

fn solution(path: &str, n_rows: u32, n_cols: u32, length: u32) -> Result<u64, Error> {
    // Read File
    let matrix: Vec<Vec<u32>> = BufReader::new(File::open(path)?)
        .lines()
        .map(|line| {
            line
                .unwrap()
                .split_whitespace()
                .filter_map(|s| s.parse().ok())
                .collect()
        })
        .collect();
         
    // Search
    let mut output = 0u64;
    for r in 0..n_rows {
        for c in 0..n_cols {
            // Vertical
            if r < n_rows - length {
                output = cmp::max(output, vertical(r, c, length, &matrix));
            }

            // Horizontal
            if c < n_cols - length {
                output = cmp::max(output, horizontal(r, c, length, &matrix));
            }

            // Diagonal "\"
            if r < n_rows - length && c < n_cols - length {
                output = cmp::max(output, diagonal_1(r, c, length, &matrix));
            }

            // Diagonal "/"
            if r < n_rows - length && c > length - 1 {
                output = cmp::max(output, diagonal_2(r, c, length, &matrix));
            }
        }
    }

    return Ok(output)
}

fn horizontal(r: u32, c: u32, n: u32, matrix: &Vec<Vec<u32>>) -> u64 {
    // Prepare
    let mut prod = 1u64;

    // Find
    for i in 0..n {
        prod *= matrix[r as usize][(c + i) as usize] as u64;
    }

    return prod
}

fn vertical(r: u32, c: u32, n: u32, matrix: &Vec<Vec<u32>>) -> u64 {
    // Prepare
    let mut prod = 1u64;

    // Find
    for i in 0..n {
        prod *= matrix[(r + i) as usize][c as usize] as u64;
    }

    return prod
}

fn diagonal_1(r: u32, c: u32, n: u32, matrix: &Vec<Vec<u32>>) -> u64 {
    // Prepare
    let mut prod = 1u64;

    // Find
    for i in 0..n {
        prod *= matrix[(r + i) as usize][(c + i) as usize] as u64;
    }

    return prod
}

fn diagonal_2(r: u32, c: u32, n: u32, matrix: &Vec<Vec<u32>>) -> u64 {
    // Prepare
    let mut prod = 1u64;

    // Find
    for i in 0..n {
        prod *= matrix[(r + i) as usize][(c - i) as usize] as u64;
    }

    return prod
}
